### PR TITLE
Allow PcapWriteHandler instances to write to the same OutputStream

### PR DESCRIPTION
**Pull request by RoganDawes**
Motivation:

Currently, PcapWriteHandler expects a dedicated OutputStream, and cannot support multiple streams written to a single pcap file. These modifications are intended to support sharing the OutputStream between multiple concurrent PcapWriteHandler instances.

Modification:

Added a "shareOutputStream" flag to PcapWriteHandler (and PcapWriter) to enable synchronization on the OutputStream in order to prevent interleaved writes. Also provide a method to write the pcap Global Header to the OutputStream.

Result:

Allows multiple channels to write to the same pcap file.

### CODE REVIEW

**Review by Kvicii — APPROVED** [review]

**Comment by RoganDawes** [comment]
On my phone, but something like:

OutputStream pcap = new FileOutputStream(pcapFile);
PcapWriteHandler.writeGlobalHeader(pcap);

Then, in the ChannelInitializer:

ch.pipeline().addFirst(new PcapWriteHandler(pcap, false, false, true));

**Comment by RoganDawes** [comment]
As an alternative, without changing the constructor signatures, simply always synchronize on the `OutputStream`. In the uncontended case, this should not add too much overhead (and to be honest, I'm pretty sure that the PcapWriteHandler is not intended for high-contention/throughput servers anyway!).

File: handler/src/main/java/io/netty/handler/pcap/PcapWriteHandler.java
Comment (hyperxpro) [inline handler/src/main/java/io/netty/handler/pcap/PcapWriteHandler.java:196]: We don't need this anymore because Pcap Global Header is already been written in `PcapWriter`.

**Review by hyperxpro — CHANGES_REQUESTED** [review]

**Comment by hyperxpro** [comment]
> synchronize

Escaping `synchronize` deadlocking can be a real thing in many scenarios. So let's leave it up to the user.

**Comment by hyperxpro** [comment]
> As an alternative, without changing the constructor signatures, simply always synchronize on the `OutputStream`. In the uncontended case, this should not add too much overhead (and to be honest, I'm pretty sure that the PcapWriteHandler is not intended for high-contention/throughput servers anyway!).

Escaping `synchronize` deadlocking can be a real thing in many scenarios. So let's leave it up to the user.

File: handler/src/main/java/io/netty/handler/pcap/PcapWriteHandler.java
Comment (RoganDawes) [reply handler/src/main/java/io/netty/handler/pcap/PcapWriteHandler.java:196]: Well, not if all the PcapWriters are constructed with the flag to NOT write the header. i.e assuming that the header has already been written, which is already an option. The problem being, who then writes that header? It could be that one PcapWriter is constructed with the flag to write the header, but then you need to be sure that this PcapWriter is the first one to actually be activated!
This method is provided to allow the header to be written in advance of any PcapWriters being constructed.

File: handler/src/main/java/io/netty/handler/pcap/PcapWriteHandler.java
Comment (daschl) [inline handler/src/main/java/io/netty/handler/pcap/PcapWriteHandler.java:195]: For consistency reasons, this might also need the checkNotNull from the constructor

File: handler/src/main/java/io/netty/handler/pcap/PcapWriter.java
Comment (daschl) [inline handler/src/main/java/io/netty/handler/pcap/PcapWriter.java:112]: An argument could be made, since the OutputStream is passed in from the outside, it should never be closed from the writer in the first place? Users might do more things with it after the handler is done with its work.

**Comment by daschl** [comment]
@RoganDawes this is great! I've been working on something similar very recently (https://review.couchbase.org/c/couchbase-jvm-clients/+/148821) and ended up running into the same thing. 👍

File: handler/src/main/java/io/netty/handler/pcap/PcapWriter.java
Comment (RoganDawes) [reply handler/src/main/java/io/netty/handler/pcap/PcapWriter.java:112]: Certainly the current PcapWriteHandler code could be used on non-concurrent instances, writing to the same OutputStream, if it were not closed. Or, one could reopen the file for appending, and not write the Global Header, and use the OutputStream to save another channel. I guess the 5.0 codebase has been forked, so these changes would need to be applied in whichever form to 5.0 as well.

**Comment by hyperxpro** [comment]
@RoganDawes Are you interested in this? https://github.com/netty/netty/issues/11543

**Comment by RoganDawes** [comment]
> @RoganDawes Are you interested in this? https://github.com/netty/netty/issues/11543

Not really part of my use case, unfortunately.

**Comment by normanmaurer** [comment]
@hyperxpro WDYT ? Can we merge this one ?

**Comment by hyperxpro** [comment]
Yup.

File: handler/src/main/java/io/netty/handler/pcap/PcapWriter.java
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/pcap/PcapWriter.java:111]:
```suggestion
            if (sharedOutputStream) {
                synchronized(outputStream) {
                    outputStream.flush();
                }
            } else {
                outputStream.flush();
                outputStream.close();
            }
```

File: handler/src/main/java/io/netty/handler/pcap/PcapWriter.java
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/pcap/PcapWriter.java:114]:
```suggestion
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
There was one missing `synchronized`

**Comment by normanmaurer** [comment]
just fixed some concurrency issue ... once the build goes green I will merge this one.

File: handler/src/main/java/io/netty/handler/pcap/PcapWriter.java
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/pcap/PcapWriter.java:70]:
```suggestion
            synchronized (outputStream) {
```

Comment (hyperxpro) [inline handler/src/main/java/io/netty/handler/pcap/PcapWriter.java:70]:
```suggestion
            synchronized (outputStream) {
```

We got to make checkstyle happy.

**Review by RoganDawes — COMMENTED** [review]
The missing synchronization is not really an issue here, because there should only ever be one thread writing the global header at the start of the file. If you have multiple threads doing that, the fact that they could stomp on each other is the least of the problems :-)

That said, I don't care strongly about that, and am happy enough that the formatting is also fixed up.

**Comment by normanmaurer** [comment]
@RoganDawes imho it is more correct to add it and cant hurt.

**Comment by normanmaurer** [comment]
Actually I think we just want to add this to the `Builder` and not add another public constructor. @RoganDawes WDYT ? If you agree, can you rebase and do the changes ?

**Comment by RoganDawes** [comment]
> Actually I think we just want to add this to the `Builder` and not add another public constructor. @RoganDawes WDYT ? If you agree, can you rebase and do the changes ?

I never even saw that there was a Builder for this. Unfortunately, I won't be in a position to make changes for a while. I guess that one would simply add a "concurrent(boolean)" method (or "sharedOutputStream(boolean)" method?), that would set the parameters appropriately, including whether to write the global header or not?

The only problem I see is that we still want to be able to write the global header at the time of construction of the OutputStream, so as to not special case the creation of each PcapWriteHandler or its Builder. i.e. if we have the "concurrent" or "shared" flag, the header should never be written, and we assume that it has already been written independently.

**Comment by hyperxpro** [comment]
@RoganDawes If you give me your authorization and permission then I'd fork your code and do fresh PR with new comments addressed.

**Comment by normanmaurer** [comment]
Let's close this... We will continue with https://github.com/netty/netty/pull/13118

**Pull request closed without merging**
